src/main/java/com/ozonehis/eip/odooopenmrs/client/OdooClient.java

```diff
@@ -0,0 +1,133 @@
+    private String username;
+
+    @Value("${odoo.password}")
+    private String password;
+
+    private Integer uid;
+
+    private XmlRpcClient client;
+
+    private static final String SERVER_OBJECT_URL = "%s/xmlrpc/2/object";
+
+    private static final String SERVER_COMMON_URL = "%s/xmlrpc/2/common";
+
+    public OdooClient() {}
+
+    public void init() throws IOException {
+        client = new XmlRpcClient() {
+            {
+                setConfig(new XmlRpcClientConfigImpl() {
+
+                    {
+                        setEnabledForExtensions(true);
+                        setServerURL(new URL(String.format(SERVER_OBJECT_URL, getUrl())));
+                    }
+                });
+            }
+        };
+    }
+
+    public void authenticate() throws MalformedURLException {
+        // TODO: Where should we place the init logic
+        try {
+            init();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        XmlRpcClientConfigImpl common_config = new XmlRpcClientConfigImpl();
+        common_config.setServerURL(new URL(String.format(SERVER_COMMON_URL, getUrl())));
+
+        try {
+            uid = (Integer) client.execute(
+                    common_config, "authenticate", asList(getDatabase(), getUsername(), getPassword(), emptyMap()));
+        } catch (XmlRpcException e) {
+            log.error("Cannot authenticate to Odoo server error: {}", e.getMessage());
+            throw new RuntimeException("Cannot authenticate to Odoo server");
+        }
+    }
+
+    public Integer create(String model, List<Map<String, Object>> dataParams)
+            throws XmlRpcException, MalformedURLException {
+        authenticateIfNecessary();
+
+        return (Integer) client.execute(
+                "execute_kw", asList(getDatabase(), uid, getPassword(), model, Constants.CREATE_METHOD, dataParams));
+    }
+
+    public Boolean write(String model, List<Object> dataParams) throws XmlRpcException, MalformedURLException {
+        authenticateIfNecessary();
+
+        return (Boolean)
+                client.execute("execute_kw", asList(getDatabase(), uid, getPassword(), model, "write", dataParams));
+    }
+
+    public Boolean delete(String model, List<Object> dataParams) throws XmlRpcException, MalformedURLException {
+        authenticateIfNecessary();
+        return (Boolean)
+                client.execute("execute_kw", asList(getDatabase(), uid, getPassword(), model, "unlink", dataParams));
+    }
+
+    public Object[] searchAndRead(String model, List<Object> criteria, List<String> fields)
+            throws XmlRpcException, MalformedURLException {
+        authenticateIfNecessary();
+
+        List<Object> params = new ArrayList<>();
+        params.add(getDatabase());
+        params.add(uid);
+        params.add(getPassword());
+        params.add(model);
+        params.add("search_read");
+        params.add(singletonList(criteria));
```

Comment:
I'm not quite sure why there's the `singletonList()` wrap here? Do you really need a `List<List<Object>>` with a single entry?

Reply:
Yes this is a how the [syntax](https://www.odoo.com/documentation/17.0/developer/reference/external_api.html#search-and-read) looks like.

![Screenshot 2024-06-11 at 12 55 02 PM](https://github.com/ozone-his/eip-odoo-openmrs/assets/36266597/2d155854-18ca-49aa-a392-198d1800a91e)

Reply:
Aha! This is why I hate XML-RPC...